server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -725,6 +786,14 @@ private static boolean assertStartedIndicesHaveCompleteTimestampRanges(ClusterSt
                         + clusterState.metadata().index(cursor.getKey()).getTimestampRange()
                         + " for "
                         + cursor.getValue().prettyPrint();
+                // assert cursor.getValue().allPrimaryShardsActive() == false
+                // || clusterState.metadata().index(cursor.getKey()).getEventIngestedRange().isComplete()
+                // : "index ["
+                // + cursor.getKey()
+                // + "] should have complete event.ingested range, but got "
+                // + clusterState.metadata().index(cursor.getKey()).getEventIngestedRange()
+                // + " for "
+                // + cursor.getValue().prettyPrint();
```

Comment:
This assert causes many of the bwc tests to fail. I've commented it out here rather than delete it, to see if anyone thinks that some form this assert should be used (and suggestions on how to do it in a backwards compatible way).

Reply:
Hmm I think if this assert doesn't hold then there's a bug somewhere. In a BwC situation the range should be `UNKNOWN` which is indeed complete.

Reply:
Got it. Thanks David.

I added some logging for when this assert fails and the event.ingested range is coming in as `NO_SHARDS` for the tests that fail.

Reply:
Since existing cluster state (of older clusters) doesn't include event.ingested range, should we instead default eventIngestedRange to UNKNOWN rather than NO_SHARDS?

I tried there: 
https://github.com/elastic/elasticsearch/pull/106252/commits/3bd5830613fa482f8d985f98e32cc7a719b830a9#diff-d9e713c7c9f70d3b97888b7ed019ba4469397004333b0d0c13fa98442beef047R1882

and with that commit all the bwc tests now pass (as well as non-bwc tests after I adjusted a few).

Reply:
Hmm I think we want to default to `NO_SHARDS` here. Instead should we treat metadata received from older nodes as if it were `UNKNOWN`?

```diff
diff --git a/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java b/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java
index 0e92b3acf12..8afee04136b 100644
--- a/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java
+++ b/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java
@@ -1780,6 +1780,8 @@ public class IndexMetadata implements Diffable<IndexMetadata>, ToXContentFragmen
         }
         if (in.getTransportVersion().onOrAfter(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
             builder.eventIngestedRange(IndexLongFieldRange.readFrom(in));
+        } else {
+            builder.eventIngestedRange(IndexLongFieldRange.UNKNOWN);
         }
         return builder.build(true);
     }
@@ -1832,6 +1834,8 @@ public class IndexMetadata implements Diffable<IndexMetadata>, ToXContentFragmen
         }
         if (out.getTransportVersion().onOrAfter(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
             eventIngestedRange.writeTo(out);
+        } else {
+            assert eventIngestedRange == IndexLongFieldRange.UNKNOWN : eventIngestedRange;
         }
     }
```

Similarly for `IndexMetadataDiff`.

I think we should also have an assertion somewhere that all these values are `UNKNOWN` until all nodes in the cluster support the new feature.